File: leetcode/pascals-triangle-ii.py
# 119. Pascal's Triangle II
# 🟢 Easy
#
# https://leetcode.com/problems/pascals-triangle-ii/
#
# Tags: Array - Dynamic Programming

import timeit
from typing import List


# Iterate over n number of rows, for each row, iterate over the row
# elements backwards computing their new value as the sum of the element
# at the same position plus the one at the previous position on the
# previous row.
#
# Time complexity: O(n^2) - We iterate n rows, for each row, we iterate
# over all the row elements.
# Space complexity: O(n) - We use an extra array of size n.
#
# Runtime 35 ms Beats 77.91%
# Memory 16.04 MB Beats 96.64%
class Solution:
    def getRow(self, rowIndex: int) -> List[int]:
        row = [0] * (rowIndex + 1)
        row[0] = 1
        for i in range(1, rowIndex + 1):
            for j in range(i, 0, -1):
                row[j] += row[j - 1]
        return row


def test():
    executors = [Solution]
    tests = [
        [0, [1]],
        [1, [1, 1]],
        [3, [1, 3, 3, 1]],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.getRow(t[0])
                exp = t[1]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/pascals-triangle-ii.rs
// 119. Pascal's Triangle II
// 🟢 Easy
//
// https://leetcode.com/problems/pascals-triangle-ii/
//
// Tags: Array - Dynamic Programming

struct Solution;
impl Solution {
    /// Generate the rows one at a time, each element is the sum of the elements
    /// at i-1 + i positions in the previous row.
    ///
    /// Time complexity: O(n^2) - We generate n rows, for each row, we visit
    /// row_length elements.
    /// Space complexity: O(n) - We use an extra vector of size n.
    ///
    /// Runtime 0 ms Beats 100%
    /// Memory 2.05 MB Beats 51.22%
    pub fn get_row(row_index: i32) -> Vec<i32> {
        let mut row = vec![1];
        let mut a = 1;
        let mut b;
        for _ in 0..row_index as usize {
            for i in 0..row.len() {
                b = a;
                a = row[i];
                if i == 0 {
                    continue;
                }
                row[i] = a + b;
            }
            row.push(1);
        }
        row
    }

    /// Similar logic but we can save the extra variable if we iterate over the
    /// elements on the row backwards.
    ///
    /// Time complexity: O(n^2) - We generate n rows, for each row, we visit
    /// row_length elements.
    /// Space complexity: O(n) - We use an extra vector of size n.
    ///
    /// Runtime 0 ms Beats 100%
    /// Memory 2.14 MB Beats 24.39%
    pub fn get_row_2(row_index: i32) -> Vec<i32> {
        let n = row_index as usize;
        let mut row = vec![0; n + 1];
        row[0] = 1;
        for i in 1..=n {
            for j in (1..=i).rev() {
                row[j] += row[j - 1];
            }
        }
        row
    }
}

// Tests.
fn main() {
    let tests = [(3, vec![1, 3, 3, 1]), (0, vec![1]), (1, vec![1, 1])];
    for t in tests {
        assert_eq!(Solution::get_row(t.0), t.1);
        assert_eq!(Solution::get_row_2(t.0), t.1);
    }
    println!("\x1b[92m» All tests passed!\x1b[0m")
}
